src/lib/tpm.c

```diff
@@ -486,6 +499,77 @@ bool files_load_bytes_from_path(const char *path, UINT8 *buf, UINT16 *size) {
 LOAD_TYPE(TPM2B_PUBLIC, public)
 LOAD_TYPE(TPM2B_PRIVATE, private)
 
+CK_RV tpm_get_token_info_from_tpm (tpm_ctx *ctx, CK_TOKEN_INFO *info) {
+
+    uint32_t revision;
+    uint32_t version;
+    uint32_t manufacturer;
+    unsigned char manufacturerID[5] = {0}; // 4 bytes + '\0'
+    uint32_t vendor[4];
+    CK_RV rv = CKR_OK;
+    TPM2_CAP capability = TPM2_CAP_TPM_PROPERTIES;
+    UINT32 property = TPM2_PT_FIXED;
+    UINT32 propertyCount = TPM2_MAX_TPM_PROPERTIES;
+    TPMS_CAPABILITY_DATA *capabilityData;
+    TPMI_YES_NO moreData;
+
+    check_pointer(ctx);
+    check_pointer(info);
+
+    TSS2_RC rval = Esys_GetCapability(ctx->esys_ctx,
+        ESYS_TR_NONE,
+        ESYS_TR_NONE,
+        ESYS_TR_NONE,
+        capability,
+        property, propertyCount, &moreData, &capabilityData);
+
+    if (rval != TSS2_RC_SUCCESS) {
+        LOGE("Esys_GetCapability: 0x%x:", rval);
+        return CKR_GENERAL_ERROR;
+    }
+
+    if (!capabilityData || capabilityData->data.tpmProperties.count < TPM2_PT_VENDOR_STRING_4 - TPM2_PT_FIXED + 1) {
+        LOGE("TPM did not reply with correct amount of capabilities");
+	rv = CKR_GENERAL_ERROR;
+	goto out;
+    }
+
+    // Use Spec revision as HW Version
+    revision = capabilityData->data.tpmProperties.tpmProperty[TPM2_PT_REVISION - TPM2_PT_FIXED].value;
+    info->hardwareVersion.major = revision / 100;
+    info->hardwareVersion.minor = revision % 100;
+
+    // Use Firmware Version as FW Version
+    version = capabilityData->data.tpmProperties.tpmProperty[TPM2_PT_FIRMWARE_VERSION_1 - TPM2_PT_FIXED].value;
+    // Most vendors seem to use 00MM.00mm as format for TPM2_PT_FIRMWARE_VERSION_1
+    // Unfortunately we only have 1 byte for major and minor each.
+    info->firmwareVersion.major = (version >> 16) & 0xFF;
+    info->firmwareVersion.minor = version  & 0xFF;
+
+    // Use Vendor String as Model
+    manufacturer = ntohl(capabilityData->data.tpmProperties.tpmProperty[TPM2_PT_MANUFACTURER - TPM2_PT_FIXED].value);
```

Comment:
so size on value should be what your expecting right? At least an assert() here would make sense. What if the data  coming back is over length ie TPM MITM?

Reply:
value is defined in the struct as an UINT32, so 4 bytes. 
These 4 bytes do NOT contain any length information, they are always 4 bytes.
If the data coming from the TPM is tampered in any way, we would read here exactly 4 byte (of garbage), but it would not harm in any way.